advanced/Scripts/dnsmasq_parse.py

```diff
@@ -0,0 +1,166 @@
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+# Pi-hole: A black hole for Internet advertisements
+# (c) 2015, 2016 by Jacob Salmela
+# Network-wide ad blocking via your Raspberry Pi
+# http://pi-hole.net
+# Controller for all pihole scripts and functions.
+#
+# Pi-hole is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+
+
+import datetime
+import sqlite3
+import time
+import sys
+import re
+
+#-----------------------------------------------------------------------------
+# Compiled Regular Expressions
+#-----------------------------------------------------------------------------
+q_re = re.compile(r'(.*) dnsmasq\[\d+\]: query\[(.*)\] (.*) from (.*)')
+f_re = re.compile(r'(.*) dnsmasq\[\d+\]: forwarded (.*) to (.*)')
+r_re = re.compile(r'(.*) dnsmasq\[\d+\]: (reply|cached) (.*) is (.*)')
+
+
+#-----------------------------------------------------------------------------
+# Functions
+#-----------------------------------------------------------------------------
+def create_tables():
+    qt = '''
+    CREATE TABLE IF NOT EXISTS queries (
+        source text,
+        query_type text,
+        name text,
+        ts datetime
+    )
+    '''
+    c.execute(qt)
+    conn.commit()
+
+    ft = '''
+    CREATE TABLE IF NOT EXISTS forwards (
+        resolver text,
+        name text,
+        ts datetime
+    )
+    '''
+    c.execute(ft)
+    conn.commit()
+
+    rt = '''
+    CREATE TABLE IF NOT EXISTS replies (
+        ip text,
+        reply_type text,
+        name text,
+        ts datetime
+    )
+    '''
+    c.execute(rt)
+    conn.commit()
+
+
+def convert_date(ds):
+    y = str(datetime.datetime.now().year)
+    ltime = time.strptime('{0} {1}'.format(y, ds), '%Y %b %d %H:%M:%S')
+
+    return time.strftime('%Y-%m-%d %H:%M:%S', ltime)
+
+
+def parse_query(query):
+    m = q_re.match(query)
+    if m is not None:
+        add_query(m.group(4), m.group(2), m.group(3), m.group(1))
+
+
+def parse_forward(query):
+    m = f_re.match(query)
+    if m is not None:
+        add_forward(m.group(3), m.group(2), m.group(1))
+
+
+def parse_reply(query):
+    m = r_re.match(query)
+    if m is not None:
+        add_reply(m.group(4), m.group(2), m.group(3), m.group(1))
+
+
+def add_query(source, qtype, name, ts):
+    sql = "INSERT INTO queries (source, query_type, name, ts) VALUES(?,?,?,?)"
+    c.execute(sql, (source, qtype, name, convert_date(ts)))
+
+
+def add_forward(resolver, name, ts):
+    sql = "INSERT INTO forwards (resolver, name, ts) VALUES(?,?,?)"
+    c.execute(sql, (resolver, name, convert_date(ts)))
+
+
+def add_reply(ip, rtype, name, ts):
+    sql = "INSERT INTO replies (ip, reply_type, name, ts) VALUES(?,?,?,?)"
```

Comment:
Specifying column names aren't necessary (although if it helps readability it's fine).